# Portfolio Project 1: Profitable App Profiles

In this project, I will demonstrate many of the basic tools and attributes that are used to perform practical data analysis on a selection, or various selections of information. My goal is to be able to master these fudamental skills useing Python 3.




In [4]:
from csv import reader

##Google Play Data Set
opened_file = open('googleplaystore.csv')
read_file = reader(opened_file)
android = list(read_file)
android_header = android[0]
android = android[1:]

## Android Play Data Set
opened_file = open('AppleStore.csv')
read_file = reader(opened_file)  #read file
ios = list(read_file)           #assign to list
ios_header = ios[0]             #assign header to first value
ios = ios[1:]                   #Assign ios to rest of data


In [6]:
def explore_data(dataset, start, end, rows_and_columns = False):
    dataset_slice = dataset[start:end]
    for row in dataset_slice:
        print(row)
        print('\n')  #adds a new (empty) line after each row
        
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

print(android_header)
print('\n')
explore_data(ios, 0, 3, True)
print('\n')
explore_data(android, 0, 3, True)

        

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7197
Number of columns: 16


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Laun

As we can see above both of the first 3 rows are listed and every row aside from the header corresponds to a particular iOS app, or an Android app (7197 apps for iOS, and 10841 apps for Android). For more information about the column information, please see this documentation on the data sets: https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home

### Deleting Wrong Data
In the discussion, there is a row in the "Google Play Data" set that has been scraped erroneously, and is providing the user with incorrect infromation about the data. Useing the information given previously, we can see the android list of lists created, as well as the header (android_header). Below we will demonstrate how to delete this row and show the difference.

In [6]:
from csv import reader

##Google Play Data Set
opened_file = open('googleplaystore.csv')
read_file = reader(opened_file)
android = list(read_file)
android_header = android[0]
android = android[1:]

print(android[10472])  # incorrect row
print('\n')
print(android_header)  # header
print('\n')
print(android[0])      # correct row

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


In [7]:
print(len(android))

10841


In [8]:
del android[10472] #Deleted Row on index 10472

In [9]:
print(len(android)) # Displays one row less..

10840


# Duplicate Entries
Sometimes, we may find duplicate entries inside of different data sets, for example, in the Google Play Data Set, we find duplicates of the Instagram app. We can use a 'for' loop to go through the data set at hand, and find duplicates.

In [12]:
for app in android:
    name  = app[0]
    if name  == 'Instagram': #If the first element is 'Instagram' display row
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


Note here, that there are 4 (four) <i>different</i> rows with duplicate apps. 

We can expand on this, and find the total amount of duplicate apps in the data by declareing two new lists, and filling them with the information found in each of the rows:

In [13]:
duplicate_apps = [] #Holds Duplicate apps in set
unique_apps = [] #Holds all unique apps in set

for app in android:
    name = app[0] #assign the app name to the first elem.
    if name in unique_apps: #if we find the duplicate, append. else - append to other list
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)

#Show number of duplicate apps, and examples of such apps:
print('Number of duplicate apps: ', len(duplicate_apps))
print('\n')
print('Examples of duplicate apps: ', duplicate_apps[:7])


Number of duplicate apps:  1181


Examples of duplicate apps:  ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits']


As we can see above, by createing two new empty lists, we are able to find all of the duplicate apps that are located within the data list with little trouble by iterating through it useing a simple for loop.

Useing the len() function on the list gives us a simple way of finding all of the duplicates.

If you examine the rows we printed two cells above for the Instagram app, the main difference happens on the fourth position of each row, which corresponds to the number of reviews. The different numbers show that the data was collected at different times. We can use this to build a criterion for keeping rows. We won't remove rows randomly, but rather we'll keep the rows that have the highest number of reviews because the higher the number of reviews, the more reliable the ratings.

To do that, we will:

*Create a dictionary where each key is a unique app name, and the value is the highest number of reviews of that app
    
*Use the dictionary to create a new data set, which will have only one entry per app (and we only select the apps with the highest number of reviews)


# Duplicate Entries (Part Two)

As mementined before to facilitate our goal in keeping rows that we want, we will create a dictionary:

In [16]:
reviews_max = {} #Empty dictionary containing maximum reviews

for row in android:
    name = row[0] #name is in the first row:
    n_reviews = float(row[3]) #num of reviews are in index 3:
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    elif name not in reviews_max: #otherwise create new entry 
        reviews_max[name] = n_reviews

#Display the number of entries (the length of the new dictionary, reviews_max:
print('The amount of entries are now: ', len(reviews_max))
    
    

The amount of entries are now:  9659


This is the expected length we wanted to have, 9659 entries.
Now, our next step is to use the same dictionary to help remove duplicate rows for the data set.


In [20]:
android_clean = [] #Stores new cleaned data set:
already_added = [] #Stores app names:

for row in android:
    name  = row[0] #Holds the name of each app
    n_reviews = float(row[3]) #Holds no. of reviews:
    if n_reviews == reviews_max[name] and (name not in already_added):
        android_clean.append(row)
        already_added.append(name)
        
#Print the data set 
print('The amount of data entries should be 9659: ', len(android_clean))


The amount of data entries should be 9659:  9659


So what did we do to obtain the data entries we wanted? 
First, we declared two new, empty lists: 'android_clean' and 'already_added'

We then iterated through the list with a 'for-loop' 
-Assigning the name to the first index in the row (row[0])
-As well as the number of reviews (n_reviews) to a casted float value in index 3 (row[3])

Useing an 'if' statement, we then checked if n_reviews had the same amount as the number of maximum reviews of the app name that is (reviews_max[name]) 'anded' if the name is not alwready in the list

If this is the case, then we will:
<ul>-Append the entire row to the list android_clean 
(via android_clean.append(row))</ul>
<ul>-Append the name of the app to the list already_added 
(via alwready_added.append(name))</ul>

<u>NOTE</u>: The reason for the if statement and the supplementary condition above is because we want the condition to account for cases where the <strong> highest number of reviews of a duplicate app is the <i>same</i> for one entry</strong>. If we just did n_reviews == reviews_max[name], we would still get duplicates later on, which isn't want we wanted.

# Removing Non-English Apps

Now that we now how to scrape the data set for duplicates, we can now focus on eliminating non-english apps, since our goal is to provide apps toward an English-speaking audience in particular. 

Below we have an example of useing both data sets, one from the Google Play Data Set and the other from the Android Play Data Set. Note that when we print some of the values from the rows, we get that is not english, but foreign.

In [22]:
from csv import reader

##Google Play Data Set
opened_file = open('googleplaystore.csv')
read_file = reader(opened_file)
android = list(read_file)
android_header = android[0]
android = android[1:]

## Android Play Data Set
opened_file = open('AppleStore.csv')
read_file = reader(opened_file)  #read file
ios = list(read_file)           #assign to list
ios_header = ios[0]             #assign header to first value
ios = ios[1:]                   #Assign ios to rest of data


print(ios[813][1])
print(ios[6731][1])
print('\n')
print(android_clean[4412][0])
print(android_clean[7940][0])

爱奇艺PPS -《欢乐颂2》电视剧热播
【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜


中国語 AQリスニング
لعبة تقدر تربح DZ


We are not interested in the following app names we have printed, so we will remove these. Each character that we use in a string corresponds to a prticular number behind the scenes. For example the character 'a' has a number 97, and the character '爱' is 29,233. We can obtain the correspoding number with a function provided by Python called ord():

Here are a few examples below:


In [23]:
print(ord('a'))
print(ord('A'))
print(ord('爱'))
print(ord('5'))
print(ord('+'))

97
65
29233
53
43


The english characters used in text all range from 0 - 127 according to the ASCII system. Based on the number system we are also able to distinguish which characters are used in english text and which are not.

In Python, strings are <i>indexable</i>, and <i>iterable</i>, which means that are are able to use indexing to select an individual character, and also iterate on the string using a 'for' loop.

Here is an example:

In [27]:
string = 'abc'

print(string[0])
print(string[1])
print(string[2])

print('\n')
print('The following should also be the same: ')
print('\n')
for character in string:
    print(character)

a
b
c


The following should also be the same: 


a
b
c


Now that we know the flexibility of Python, lets write a function that will tells us if the string passed as an argument contains characters that are not found in English text:

In [36]:
def notEngChar(string_):
    for char in string_:
        if ord(char) > 127:
            return True
    return False


#Display the result:
print(notEngChar('Instagram'))
print(notEngChar('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(notEngChar('Docs To Go™ Free Office Suite'))
print(notEngChar('Instachat 😜'))

False
True
True
True


Note that we can also assign the string to a variable, and 
our result will be the same:

In [37]:
#This will yeild the same result:
str1 = 'Instachat 😜'
print(notEngChar(str1))

True


# Removing Non-English Apps: Part 2

Previously, we used an algorithm to find out which string contained characters not found in English text. However, we saw that the function we wrote had a coule of mistakes, as we saw when we used a string like 'Instachat 😜' or 'Docs To Go™ Free Office Suite'.

This is due to the fact that some emojis fall out-of-range from our check. 

We can see this by printing the following characters:

In [38]:
print(ord('™'))
print(ord('😜'))


8482
128540


Instead of just checking one character, another option would be checking three characters or less that fall outside the ASCII range:

In [49]:
def explore_data(dataset, start, end, rows_and_columns = False):
    dataset_slice = dataset[start:end]
    for row in dataset_slice:
        print(row)
        print('\n')  #adds a new (empty) line after each row
        
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))




def ThreeChars(string_):
    charCount = 0;
    #Use a for loop to iterate through the string,
    #if there is a character beyond 127, increment charCount.
    for char in string_:
        if (ord(char) > 127):
            charCount+=1
    #If the number of char. exceeds 3 return false, otherwise return true
    if charCount > 3:
        return False
    else:
        return True


    
#Some examples of output:
print(ThreeChars('Instachat 😜'))
print(ThreeChars('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(ThreeChars('Docs To Go™ Free Office Suite'))

android_english = []
ios_english = []

for app in android_clean:
    name = app[0]
    if ThreeChars(name):
        android_english.append(app)
        
for app in ios:
    name = app[1]
    if ThreeChars(name):
        ios_english.append(app)
        
explore_data(android_english, 0, 3, True)
print('\n')
explore_data(ios_english, 0, 3, True)


True
False
True
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '

# Isolating Free Apps

As I mentoned in the intro, the company only builds apps that are free to download and install, and the main source of revenue behind this consists of in-app advertisements. These data sets contain a combination of free and non free apps. We can then isolate these from the data set by creating a new list, and fill it with free 

In [51]:

#Create new empty, lists:
android_apps_final = []
ios_apps_final = []

for app in android_english:
    price = app[7]
    if price == '0':
        android_apps_final.append(app)

for app in ios_english:
    price = app[4]
    if price == '0.0':
        ios_apps_final.append(app)
        
        
#Check length of each data set:
print(len(android_apps_final))
print(len(ios_apps_final))


8864
3222


We are then left with 8864 android apps, and 3222 iOS apps that remain.

# Most Common Apps by Genre: Part One
As we mentioned in the introduction, our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps.

To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:

<ul>Build a minimal Android version of the app, and add it to Google Play.</ul>
<ul>If the app has a good response from users, we then develop it further.</ul>
<ul>If the app is profitable after six months, we also build an iOS version of the app and add it to the App Store.</ul>

Because our end goal is to add the app on both the App Store and Google Play, we need to find app profiles that are successful on both markets. For instance, a profile that might work well for both markets might be a productivity app that makes use of gamification.

Let's begin the analysis by getting a sense of the most common genres for each market. For this, we'll build a frequency table for the prime_genre column of the App Store data set, and the Genres and Category columns of the Google Play data set.

# Part Two

We'll build two functions we can use to analyze the frequency tables:

<ul>One function to generate frequency tables that show percentages.</ul>
<ul>Another function that we can use to display the percentages in a descending order</ul>


In [56]:
def freq_table(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total+=1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
            
    table_percentages = {}
    
    for key in table:
        percentage = (table[key] / total) * 100
        table_percentages[key] = percentage
        
    return table_percentages

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])
    
    
    

# Part Three
We start by examining the frequency table for the prime_genre column of the App Store data set.

In [57]:
#Freq. Table for prime_genre column in App Store:
display_table(ios_apps_final, -5)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665



Above, we can see that among the free English apps, more than a half (58.16%) are games. Entertainment apps are close to 8%, followed by photo and video apps, which are close to 5%. Only 3.66% of the apps are designed for education, followed by social networking apps which amount for 3.29% of the apps in our data set.

The general impression is that App Store (at least the part containing free English apps) is dominated by apps that are designed for fun (games, entertainment, photo and video, social networking, sports, music, etc.), while apps with practical purposes (education, shopping, utilities, productivity, lifestyle, etc.) are more rare. However, the fact that fun apps are the most numerous doesn't also imply that they also have the greatest number of users — the demand might not be the same as the offer.

Let's continue by examining the Genres and Category columns of the Google Play data set (two columns which seem to be related).


In [60]:
display_table(android_apps_final, 1)

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 



The landscape seems significantly different on Google Play: there are not that many apps designed for fun, and it seems that a good number of apps are designed for practical purposes (family, tools, business, lifestyle, productivity, etc.). However, if we investigate this further, we can see that the family category (which accounts for almost 19% of the apps) means mostly games for kids.

Even so, practical apps seem to have a better representation on Google Play compared to App Store. This picture is also confirmed by the frequency table we see for the Genres column:


In [61]:
display_table(android_apps_final, -4)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

The difference between the Genres and the Category columns is not crystal clear, but one thing we can notice is that the Genres column is much more granular (it has more categories). We're only looking for the bigger picture at the moment, so we'll only work with the Category column moving forward.

Up to this point, we found that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. Now we'd like to get an idea about the kind of apps that have most users.

# Most Popular Apps by Genre on the App Store
One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. 

In [65]:
genres_ios = freq_table(ios_apps_final, -5)

for genre in genres_ios:
    total = 0
    len_genre = 0
    for app in ios_apps_final:
        genre_app = app[-5]
        if genre_app == genre:            
            n_ratings = float(app[5])
            total += n_ratings
            len_genre += 1
    avg_n_ratings = total / len_genre
    print(genre, ':', avg_n_ratings)

Music : 57326.530303030304
Education : 7003.983050847458
Travel : 28243.8
Business : 7491.117647058823
Sports : 23008.898550724636
Finance : 31467.944444444445
Photo & Video : 28441.54375
Utilities : 18684.456790123455
Reference : 74942.11111111111
Entertainment : 14029.830708661417
Weather : 52279.892857142855
Social Networking : 71548.34905660378
Productivity : 21028.410714285714
Navigation : 86090.33333333333
Shopping : 26919.690476190477
Food & Drink : 33333.92307692308
Health & Fitness : 23298.015384615384
News : 21248.023255813954
Lifestyle : 16485.764705882353
Games : 22788.6696905016
Book : 39758.5
Catalogs : 4004.0
Medical : 612.0


In [67]:
for app in ios_apps_final:
    if app[-5] == 'Navigation':
        print(app[1], ':', app[5]) # print name and number of ratings

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5




The same pattern applies to social networking apps, where the average number is heavily influenced by a few giants like Facebook, Pinterest, Skype, etc. Same applies to music apps, where a few big players like Pandora, Spotify, and Shazam heavily influence the average number.

Our aim is to find popular genres, but navigation, social networking or music apps might seem more popular than they really are. The average number of ratings seem to be skewed by very few apps which have hundreds of thousands of user ratings, while the other apps may struggle to get past the 10,000 threshold. We could get a better picture by removing these extremely popular apps for each genre and then rework the averages, but we'll leave this level of detail for later.

Reference apps have 74,942 user ratings on average, but it's actually the Bible and Dictionary.com which skew up the average rating:


In [69]:
for app in ios_apps_final:
    if app[-5] == 'Reference':
        print(app[1], ':', app[5])


Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0




However, this niche seems to show some potential. One thing we could do is take another popular book and turn it into an app where we could add different features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes about the book, etc. On top of that, we could also embed a dictionary within the app, so users don't need to exit our app to look up words in an external app.

This idea seems to fit well with the fact that the App Store is dominated by for-fun apps. This suggests the market might be a bit saturated with for-fun apps, which means a practical app might have more of a chance to stand out among the huge number of apps on the App Store.

Other genres that seem popular include weather, book, food and drink, or finance. The book genre seem to overlap a bit with the app idea we described above, but the other genres don't seem too interesting to us:

<ul>Weather apps — people generally don't spend too much time in-app, and the chances of making profit from in-app adds are low. Also, getting reliable live weather data may require us to connect our apps to non-free APIs.</ul>

<ul>Food and drink — examples here include Starbucks, Dunkin' Donuts, McDonald's, etc. So making a popular food and drink app requires actual cooking and a delivery service, which is outside the scope of our company.</ul>

<ul>Finance apps — these apps involve banking, paying bills, money transfer, etc. Building a finance app requires domain knowledge, and we don't want to hire a finance expert just to build an app.</ul>

Now let's analyze the Google Play market a bit.

# Most Popular Apps on Google Play
For the Google Play market, we actually have data about the number of installs, so we should be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.):

In [70]:
display_table(android_apps_final, 5) # the Installs columns

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343




One problem with this data is that is not precise. For instance, we don't know whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don't need very precise data for our purposes — we only want to get an idea which app genres attract the most users, and we don't need perfect precision with respect to the number of users.

We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on.

To perform computations, however, we'll need to convert each install number to float — this means that we need to remove the commas and the plus characters, otherwise the conversion will fail and raise an error. We'll do this directly in the loop below, where we also compute the average number of installs for each genre (category).


In [73]:


categories_android = freq_table(android_apps_final, 1)

for category in categories_android:
    total = 0
    len_category = 0
    for app in android_apps_final:
        category_app = app[1]
        if category_app == category:            
            n_installs = app[5]
            n_installs = n_installs.replace(',', '')
            n_installs = n_installs.replace('+', '')
            total += float(n_installs)
            len_category += 1
    avg_n_installs = total / len_category
    print(category, ':', avg_n_installs)



SOCIAL : 23253652.127118643
HEALTH_AND_FITNESS : 4188821.9853479853
DATING : 854028.8303030303
WEATHER : 5074486.197183099
HOUSE_AND_HOME : 1331540.5616438356
TOOLS : 10801391.298666667
SPORTS : 3638640.1428571427
BUSINESS : 1712290.1474201474
BOOKS_AND_REFERENCE : 8767811.894736841
COMICS : 817657.2727272727
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
LIFESTYLE : 1437816.2687861272
MEDICAL : 120550.61980830671
PERSONALIZATION : 5201482.6122448975
EVENTS : 253542.22222222222
FOOD_AND_DRINK : 1924897.7363636363
LIBRARIES_AND_DEMO : 638503.734939759
PHOTOGRAPHY : 17840110.40229885
ENTERTAINMENT : 11640705.88235294
GAME : 15588015.603248259
PRODUCTIVITY : 16787331.344927534
FAMILY : 3695641.8198090694
TRAVEL_AND_LOCAL : 13984077.710144928
EDUCATION : 1833495.145631068
SHOPPING : 7036877.311557789
FINANCE : 1387692.475609756
PARENTING : 542603.6206896552
COMMUNICATION : 38456119.167247385
NEWS_AND_MAGAZINES : 9549178.467741935
VIDEO_PLAYERS : 24727872.452830188
MAPS_A



On average, communication apps have the most installs: 38,456,119. This number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts), and a few others with over 100 and 500 million installs:


In [75]:
for app in android_apps_final:
    if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+'
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])



WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess



If we removed all the communication apps that have over 100 million installs, the average would be reduced roughly ten times:


In [76]:
under_100_m = []

for app in android_apps_final:
    n_installs = app[5]
    n_installs = n_installs.replace(',', '')
    n_installs = n_installs.replace('+', '')
    if (app[1] == 'COMMUNICATION') and (float(n_installs) < 100000000):
        under_100_m.append(float(n_installs))
        
sum(under_100_m) / len(under_100_m)

3603485.3884615386



We see the same pattern for the video players category, which is the runner-up with 24,727,872 installs. The market is dominated by apps like Youtube, Google Play Movies & TV, or MX Player. The pattern is repeated for social apps (where we have giants like Facebook, Instagram, Google+, etc.), photography apps (Google Photos and other popular photo editors), or productivity apps (Microsoft Word, Dropbox, Google Calendar, Evernote, etc.).

Again, the main concern is that these app genres might seem more popular than they really are. Moreover, these niches seem to be dominated by a few giants who are hard to compete against.

The game genre seems pretty popular, but previously we found out this part of the market seems a bit saturated, so we'd like to come up with a different app recommendation if possible.

The books and reference genre looks fairly popular as well, with an average number of installs of 8,767,811. It's interesting to explore this in more depth, since we found this genre has some potential to work well on the App Store, and our aim is to recommend an app genre that shows potential for being profitable on both the App Store and Google Play.

Let's take a look at some of the apps from this genre and their number of installs:


In [78]:
for app in android_apps_final:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(app[0], ':', app[5])

E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
ReadEra – free ebook reader : 1,000,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
E

This niche seems to be dominated by software for processing and reading ebooks, as well as various collections of libraries and dictionaries, so it's probably not a good idea to build similar apps since there'll be some significant competition.

We also notice there are quite a few apps built around the book Quran, which suggests that building an app around a popular book can be profitable. It seems that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets.

However, it looks like the market is already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.

# Conclusion
In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

We concluded that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets. The markets are already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.